src/EntityId/PrefixingEntityIdParser.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Wikibase\DataModel\Services\EntityId;
+
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+
+/**
+ * EntityIdParser that adds a fixed prefix and parses resulting string as an EntityId.
+ * This can be used to prefix IDs of entities coming from a foreign repository
+ * with the repository name to avoid clashes with IDs of local entities.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class PrefixingEntityIdParser implements EntityIdParser {
+
+	/**
+	 * @var string
+	 */
+	private $prefix;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $idParser;
+
+	/**
+	 * @param string $prefix Prefix to be added
+	 * @param EntityIdParser $idParser
+	 */
+	public function __construct( $prefix, EntityIdParser $idParser ) {
+		$this->prefix = $prefix;
+		$this->idParser = $idParser;
+	}
+
+	/**
+	 * Adds the prefix to the given string and parses the resulting string as an EntityId.
+	 * If the string is already prefixed, the prefix is not duplicated (this check is case sensitive).
+	 *
+	 * @param string $idSerialization
+	 * @return EntityId
+	 */
+	public function parse($idSerialization) {
+		// Do not add prefix if $idSerialization already starts with it
```

Comment:
Wait, why does a parser *add* something that was not given in the input? This feels wrong. This means: All inputs with a prefix not supported by this parser will "successfully" be parsed. Or in other words: This parser assumes all inputs either start with the expected or none prefix. We would need some kind of PrefixValidator to make sure this is always the case. I don't think this is the best approach.